Add optional default value to average template function/filter

ehendrix23 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Return optional default value on empty list or non-numeric characters in lust.

If an empty list is provided to the average function it will currently fail, this change allows one to provide a default value to be returned instead.
For example, getting the average of 2 state objects:
```
{{ [float(states('sensor.openweathermap_feels_like_temperature'),'NA'),float(state_attr('weather.main_floor','temperature'),'NA')] | select('is_number') | average }}
```
If both are unavailable or not returning a number then the list is empty resulting in an exception. With this change a default value can be provided instead. So one could then do for example:
```
{{ [float(states('sensor.openweathermap_feels_like_temperature'),'NA'),float(state_attr('weather.main_floor','temperature'),'NA')] | select('is_number') | average(default='unavailable') }}
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24282

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ehendrix23 committed:
Return None on empty list

ehendrix23 commented:
Note, another solution for this could be to check if the 1st argument is an iterable and if it is then use 2nd argument for default.
One has to pass a list: ```average([1,2])``` or multiple arguments: ```average(1,2)```. Passing a list with multiple arguments ```average([1,2],2)``` returns an error now. Hence we could make it that if ```average([1,2],2)``` is passed that 2 is then the default in case the list passed is empty. This would be a bit more in line with some of the other functions where there is a default.

So using my earlier example it would then be:
```{{ [float(states('sensor.openweathermap_feels_like_temperature'),'NA'),float(state_attr('weather.main_floor','temperature'),'NA')] | select('is_number') | average('unavailable') }}```

frenck commented:
I'm looking at this, and doubting if this needs the default argument or not. 

@emontnemery What do you think?

emontnemery commented:
I think that the implementation should be updated to allow passing a default argument (for example `None`), and if no default argument is passed the function/filter still throws. That's more in line with how other template functions work when they get invalid data.

An alternative could be:
- Mark the PR as a breaking change
- The implementation should be updated to allow passing a default argument other than `None`

emontnemery commented:
@ehendrix23 templates support named arguments, meaning we can handle both cases:

If the 1st argument is a list, the default is the 2nd argument or a named argument `default`
If the 1st argument is not a list, the default is a named argument `default`

We handle this case in the `atan2` filter/function already: https://github.com/home-assistant/core/blob/fab3ee90b2dbcc553297716515c95ab7331fcee7/homeassistant/helpers/template.py#L1510

ehendrix23 committed:
Updated to use default value

ehendrix23 commented:
Updated it based on recommended approach. Let me know how this looks.
I've added tests accordingly as well which all passed.

Also, when trying to commit I error that pylint failed:
```homeassistant/helpers/template.py:203:4: W0231: __init__ method from base class 'tuple' is not called (super-init-not-called)```

I did not even touch that area so not sure why it would be failing in this now. I circumvented it with adding:
```# pylint: disable=super-init-not-called``` just before that definition but not sure if this is right.

emontnemery commented:
I don't see any pylint errors when removing the `disable=super-init-not-called`, is your environment updated?

I get:
```
$ pylint --version
pylint 2.15.0
astroid 2.12.5
Python 3.10.4 (main, Aug 31 2022, 08:30:28) [GCC 9.3.0]
```

emontnemery commented on homeassistant/helpers/template.py:
```diff
@@ -1658,13 +1659,23 @@ def average(*args: Any) -> float:
-        if isinstance(args[0], Iterable):
-            return statistics.fmean(args[0])
-
+    # If first argument is iterable, more then 1 argument provided but not named default
+    # than use 2nd argument as default.
```
```suggestion
    # If first argument is an iterable and more than 1 argument provided but not a named default,
    # then use the 2nd argument as default.
```

emontnemery reviewed (commented):
Looks good!

Small suggestion to improve grammar of a comment + please double check if CI passes without the pylint override.

ehendrix23 committed:
Update comment.

ehendrix23 commented:
Updating pylint fixed that issue. I was under the impression as I run the Home assistant development docker it would keep those things up-to-date. Clearly I was mistaken. :-)

Updated the comment.

emontnemery reviewed (approved):
Thanks, @ehendrix23 👍
